src/Shared/test/MockTimeProvider.cs

```diff
@@ -0,0 +1,88 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+
+namespace Microsoft.AspNetCore.Testing;
+
+public class MockTimeProvider : TimeProvider
+{
+    private readonly long _timestampFrequency =
+        // 10_000_000; // Windows
+        // 100_000_000; // 8 zeros, in between
+        // 1_000_000_000; // Linux
+        Stopwatch.Frequency;
+    private long _utcTicks;
+    private long _timestamp;
+
+    public MockTimeProvider()
+    {
+        // Use a random DateTimeOffset to ensure tests that incorrectly use the current DateTimeOffset fail always instead of only rarely.
+        var tenYears = TimeSpan.FromDays(365 * 10).Ticks;
+        _utcTicks = DateTimeOffset.UtcNow.Ticks + Random.Shared.NextInt64(-tenYears, tenYears);
+        // Timestamps often measure system uptime.
+        _timestamp = Random.Shared.NextInt64(0, System.GetTimestamp() * 100);
+    }
+
+    public MockTimeProvider(DateTimeOffset now)
+    {
+        _utcTicks = now.Ticks;
+        // Timestamps often measure system uptime.
+        _timestamp = Random.Shared.NextInt64(0, System.GetTimestamp() * 100);
+    }
+
+    public override DateTimeOffset GetUtcNow()
+    {
+        UtcNowCalled++;
+        return new DateTimeOffset(Interlocked.Read(ref _utcTicks), TimeSpan.Zero);
+    }
+
+    public override long GetTimestamp() => Interlocked.Read(ref _timestamp);
+
+    public override long TimestampFrequency => _timestampFrequency;
+
+    public int UtcNowCalled { get; private set; }
+
+    public void Advance(TimeSpan timeSpan)
+    {
+        Interlocked.Add(ref _utcTicks, timeSpan.Ticks);
+        Interlocked.Add(ref _timestamp, timeSpan.Ticks * (_timestampFrequency / TimeSpan.TicksPerSecond));
+
+        if (_utcTicks < 0 || _timestamp < 0)
+        {
+            throw new InvalidOperationException("UtcNow or Timestamp became negative.");
+        }
+    }
+
+    public void AdvanceTo(DateTimeOffset newUtcNow)
+    {
+        var nowTicks = newUtcNow.UtcTicks;
+        var priorTicks = Interlocked.Exchange(ref _utcTicks, nowTicks);
+        // Advance Timestamp by the same amount.
+        // Known timestamp frequencies are the same or larger than TicksPerSecond.
+        var timestampOffset = (nowTicks - priorTicks) * (_timestampFrequency / TimeSpan.TicksPerSecond);
+        Interlocked.Add(ref _timestamp, timestampOffset);
+
+        if (_utcTicks < 0 || _timestamp < 0)
```

Comment:
I know this is just test code, but why not verify that `timestampOffset` is positive? Do we have a test that intentionally moves the clock backwards? I think this that's what @BrennanConroy was proposing.

```suggestion
        if (timestampOffset < 0 || _utcTicks < 0 || _timestamp < 0)
```